nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/PutElasticsearchJson.java

```diff
@@ -0,0 +1,424 @@
+                "and this option is used for getting the ID, an exception will be raised.")
+        .required(false)
+        .expressionLanguageSupported(ExpressionLanguageScope.FLOWFILE_ATTRIBUTES)
+        .addValidator(Validator.VALID)
+        .build();
+    static final PropertyDescriptor ID_JSON_PATH = new PropertyDescriptor.Builder()
+        .name("put-es-json-id-json-path")
+        .displayName("ID JSONPath")
+        .description("If set, the document ID will be pulled from each JSON block using this JSONPath operation.")
+        .required(false)
+        .expressionLanguageSupported(ExpressionLanguageScope.FLOWFILE_ATTRIBUTES)
+        .addValidator(Validator.VALID)
+        .build();
+    static final PropertyDescriptor INDEX_JSON_PATH = new PropertyDescriptor.Builder()
+        .name("put-es-json-index-json-path")
+        .displayName("Index JSONPath")
+        .description("If set, the document's target index will be pulled from each JSON block using this JSONPath operation. " +
+                "If it is left blank or fails to evaluate to a valid value, the index property's valuable will be used instead. ")
+        .required(false)
+        .expressionLanguageSupported(ExpressionLanguageScope.FLOWFILE_ATTRIBUTES)
+        .addValidator(Validator.VALID)
+        .build();
+    static final PropertyDescriptor TYPE_JSON_PATH = new PropertyDescriptor.Builder()
+        .name("put-es-json-type-json-path")
+        .displayName("Type JSONPath")
+        .description("If set, the document's target type will be pulled from each JSON block using this JSONPath operation. " +
+                "If it is left blank or fails to evaluate to a valid value, the type property's valuable will be used instead. ")
+        .required(false)
+        .expressionLanguageSupported(ExpressionLanguageScope.FLOWFILE_ATTRIBUTES)
+        .addValidator(Validator.VALID)
+        .build();
+    static final PropertyDescriptor BATCH_SIZE = new PropertyDescriptor.Builder()
+        .name("put-es-json-batch-size")
+        .displayName("Batch Size")
+        .description("The number of flowfiles to fetch per batch. Be very careful about setting this value high if" +
+            "JSON arrays could be sent in the flowfiles.")
+        .required(true)
+        .defaultValue("50")
+        .expressionLanguageSupported(ExpressionLanguageScope.VARIABLE_REGISTRY)
+        .addValidator(StandardValidators.POSITIVE_INTEGER_VALIDATOR)
+        .build();
+
+    static final PropertyDescriptor SPLIT_OFF_FAILED_RECORDS = new PropertyDescriptor.Builder()
+        .name("put-es-json-split-off-failed-records")
+        .displayName("Split off Failed Records")
+        .displayName("If this is enabled, the response from Elasticsearch will be examined for failure messages and failed " +
+                "json records will be sent to the errors relationship.")
```

Comment:
How are these output? Since there's no RecordWriter I'm assuming they'd come out as a top-level JSON array? Are successful records written out the same way? The original input flowfile could have the records as a nested array, so it'd be helpful to know what is written out where. When other processors can split the output, there's usually an `original` relationship for example.

Reply:
Same.